internal/providercache/dir_modify.go

```diff
@@ -0,0 +1,121 @@
+		return installFromLocalArchive(ctx, string(location), newPath)
+	case getproviders.PackageLocalDir:
+		return installFromLocalDir(ctx, string(location), newPath)
+	default:
+		// Should not get here, because the above should be exhaustive for
+		// all implementations of getproviders.Location.
+		return fmt.Errorf("don't know how to install from a %T location", location)
+	}
+}
+
+// LinkFromOtherCache takes a CachedProvider value produced from another Dir
+// and links it into the cache represented by the receiver Dir.
+//
+// This is used to implement tiered caching, where new providers are first
+// populated into a system-wide shared cache and then linked from there into
+// a configuration-specific local cache.
+//
+// It's invalid to link a CachedProvider from a particular Dir into that same
+// Dir, because that would otherwise potentially replace a real package
+// directory with a circular link back to itself.
+func (d *Dir) LinkFromOtherCache(entry *CachedProvider) error {
+	newPath := getproviders.UnpackedDirectoryPathForPackage(
+		d.baseDir, entry.Provider, entry.Version, d.targetPlatform,
+	)
+	currentPath := entry.PackageDir
+
+	absNew, err := filepath.Abs(newPath)
+	if err != nil {
+		return fmt.Errorf("failed to make new path %s absolute: %s", newPath, err)
+	}
+	absCurrent, err := filepath.Abs(currentPath)
+	if err != nil {
+		return fmt.Errorf("failed to make existing cache path %s absolute: %s", currentPath, err)
+	}
+
+	// Before we do anything else, we'll do a quick check to make sure that
+	// these two paths are not pointing at the same physical directory on
+	// disk. This compares the files by their OS-level device and directory
+	// entry identifiers, not by their virtual filesystem paths.
+	if same, err := copydir.SameFile(absNew, absCurrent); same {
+		return fmt.Errorf("cannot link existing cache path %s to itself", newPath)
+	} else if err != nil {
+		return fmt.Errorf("failed to determine if %s and %s are the same: %s", currentPath, newPath, err)
+	}
+
+	// Invalidate our metaCache so that subsequent read calls will re-scan to
+	// incorporate any changes we make here.
+	d.metaCache = nil
+
+	// Delete anything that's already present at this path first.
+	err = os.RemoveAll(currentPath)
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("failed to remove existing %s before linking it to %s: %s", currentPath, newPath, err)
+	}
+
+	// We'll prefer to create a symlink if possible, but we'll fall back to
+	// a recursive copy if symlink creation fails. It could fail for a number
+	// of reasons, including being on Windows 8 without administrator
+	// privileges or being on a legacy filesystem like FAT that has no way
+	// to represent a symlink. (Generalized symlink support for Windows was
+	// introduced in a Windows 10 minor update.)
+	//
+	// We'd prefer to use a relative path for the symlink to reduce the risk
+	// of it being broken by moving things around later, but we'll fall back
```

Comment:
I'd expect a relative path to be more likely to break than an absolute one. I'd be surprised if the configuration directory is moved along with any global cache directory—I'd expect configurations to move around, and the global cache to stay still. Do I misunderstand?

Reply:
This is a fair point... I will be honest and say that I just mimicked what our old cache codepath was doing here without thinking too critically about it. I'll switch this to be absolute in a later PR.